### PR TITLE
[EVM] Support commutable operands in the stackification algorithm

### PR DESCRIPTION
**Pull request by PavelKopyl**
# Code Review Checklist

## Requirements
- [ ] Have the requirements been met?
- [ ] Have stakeholder(s) approved the change?

## Implementation
- [ ] Does this code change accomplish what it is supposed to do?
- [ ] Can this solution be simplified?
- [ ] Does this change add unwanted compile-time or run-time dependencies?
- [ ] Could an additional framework, API, library, or service improve the solution?
- [ ] Could we reuse part of LLVM instead of implementing the patch or a part of it?
- [ ] Is the code at the right abstraction level?
- [ ] Is the code modular enough?
- [ ] Can a better solution be found in terms of maintainability, readability, performance, or security?
- [ ] Does similar functionality already exist in the codebase? If yes, why isn’t it reused?
- [ ] Are there any best practices, design patterns or language-specific patterns that could substantially improve this code? 

## Logic Errors and Bugs
- [ ] Can you think of any use case in which the
code does not behave as intended?
- [ ] Can you think of any inputs or external events
that could break the code?

## Error Handling and Logging
- [ ] Is error handling done the correct way?
- [ ] Should any logging or debugging information
be added or removed?
- [ ] Are error messages user-friendly?
- [ ] Are there enough log events and are they
written in a way that allows for easy
debugging?

## Maintainability
- [ ] Is the code easy to read?
- [ ] Is the code not repeated (DRY Principle)?
- [ ] Is the code method/class not too long?

## Dependencies
- [ ] Were updates to documentation, configuration, or readme files made as required by this change?
- [ ] Are there any potential impacts on other parts of the system or backward compatibility?

## Security
- [ ] Does the code introduce any security vulnerabilities?

## Performance
- [ ] Do you think this code change decreases
system performance?
- [ ] Do you see any potential to improve the
performance of the code significantly?

## Testing and Testability
- [ ] Is the code testable?
- [ ] Have automated tests been added, or have related ones been updated to cover the change?
	- [ ] For changes to mutable state
- [ ] Do tests reasonably cover the code change (unit/integration/system tests)? 
	- [ ] Line Coverage
	- [ ] Region Coverage
	- [ ] Branch Coverage
- [ ] Are there some test cases, input or edge cases
that should be tested in addition?

## Readability
- [ ] Is the code easy to understand?
- [ ] Which parts were confusing to you and why?
- [ ] Can the readability of the code be improved by
smaller methods?
- [ ] Can the readability of the code be improved by
different function, method or variable names?
- [ ] Is the code located in the right
file/folder/package?
- [ ] Do you think certain methods should be
restructured to have a more intuitive control
flow?
- [ ] Is the data flow understandable?
- [ ] Are there redundant or outdated comments?
- [ ] Could some comments convey the message
better?
- [ ] Would more comments make the code more
understandable?
- [ ] Could some comments be removed by making the code itself more readable?
- [ ] Is there any commented-out code?
- [ ] Have you run a spelling and grammar checker?

## Documentation
- [ ] Is there sufficient documentation?
- [ ] Is the ReadMe.md file up to date?

## Best Practices
- [ ] Follow Single Responsibility principle?
- [ ] Are different errors handled correctly?
- [ ] Are errors and warnings logged?
- [ ] Magic values avoided?
- [ ] No unnecessary comments?
- [ ] Minimal nesting used?

## Experts' Opinion
- [ ] Do you think a specific expert, like a security
expert or a usability expert, should look over
the code before it can be accepted?
- [ ] Will this code change impact different teams, and should they review the change as well?

### CODE REVIEW

**Comment by PavelKopyl** [comment]
I haven't yet find relatively short examples to demonstrate the effect. All of them star from ~300 asm lines. Will later try to reduce them.

**Comment by akiramenai** [comment]
@PavelKopyl it doesn't improve https://github.com/matter-labs/era-compiler-llvm/pull/714/files#diff-eedcb7abb5e457034f724cbac18440f6320c12176cea91aca40c608c17030e9e, does it?

File: llvm/lib/Target/EVM/EVMControlFlowGraph.h [hunk 1]
```diff
@@ -168,6 +168,12 @@ struct CFG {
 
   struct BuiltinCall {
     MachineInstr *Builtin = nullptr;
+    // This contains commutable operand indexes, if any.
+    // Operand indexes correspond to a stack layout, i.e., the first use
+    // operand in the instruction definition is located on the stack top
+    // and has zero index.
+    std::optional<std::pair<unsigned, unsigned>> CommutableOpIndexes =
```
Comment (akiramenai) [inline llvm/lib/Target/EVM/EVMControlFlowGraph.h]: It is set to `{0, 1}` in `ControlFlowGraphBuilder::getCommutableOpIndexes` but it's never updated. Can't we use a flag instead?

**Comment by PavelKopyl** [comment]
> @PavelKopyl it doesn't improve https://github.com/matter-labs/era-compiler-llvm/pull/714/files#diff-eedcb7abb5e457034f724cbac18440f6320c12176cea91aca40c608c17030e9e, does it?

I've slightly extended commutable operations support to take into account stack layout formed by the function arguments. But, not all the cases are covered, especially when it's cheaper to get rid of JUNK slots later, not before the first instruction in the MF's entry block. The limitation is related to the local nature of the stackification algorithm. We need to teach it somehow to take into account stack transformations between several operations not just between two adjacent.

File: llvm/lib/Target/EVM/EVMControlFlowGraph.h (hunk 1 above)
Comment (PavelKopyl) [reply llvm/lib/Target/EVM/EVMControlFlowGraph.h]: Sure we can, I just initially wanted to make it a bit more generalized.

File: llvm/lib/Target/EVM/EVMControlFlowGraphBuilder.cpp [hunk 2]
```diff
@@ -182,26 +184,40 @@ StackSlot ControlFlowGraphBuilder::getDefiningSlot(const MachineInstr &MI,
...
+  }
+}
+
+bool ControlFlowGraphBuilder::isCommutable(const MachineInstr &MI) const {
+  // In all the cases, commutable operands take top two stack slots.
+  // Even for ADDMOD and MULMOD instructions, because their MOD operand
+  // takes third slot counted from the top of stack.
+  unsigned Opc = MI.getOpcode();
+  return (MI.isCommutable() || Opc == EVM::ADDMOD || Opc == EVM::MULMOD)
+             ? true
```
Comment (akiramenai) [inline llvm/lib/Target/EVM/EVMControlFlowGraphBuilder.cpp]: Redundant ternary

File: llvm/lib/Target/EVM/EVMOptimizedCodeTransform.cpp [hunk 3]
```diff
@@ -122,12 +122,17 @@ EVMOptimizedCodeTransform::EVMOptimizedCodeTransform(EVMAssembly &Assembly,
...
     : Assembly(Assembly), Layout(Layout), FuncInfo(&Cfg.FuncInfo), MF(MF) {}
 
-void EVMOptimizedCodeTransform::assertLayoutCompatibility(
-    Stack const &SourceStack, Stack const &TargetStack) {
-  assert(SourceStack.size() == TargetStack.size());
+bool EVMOptimizedCodeTransform::AreLayoutsCompatible(Stack const &SourceStack,
+                                                     Stack const &TargetStack) {
+  if (SourceStack.size() != TargetStack.size())
+    return false;
+
```
Comment (akiramenai) [inline llvm/lib/Target/EVM/EVMOptimizedCodeTransform.cpp:130]: The remainder can be folded to an algorithm (all / any)

File: llvm/test/CodeGen/EVM/commutable-ops.ll [hunk 4]
```diff
@@ -0,0 +1,39 @@
+; RUN: llc < %s | FileCheck %s
```
Comment (akiramenai) [inline llvm/test/CodeGen/EVM/commutable-ops.ll]: Perhaps worth merging with `stack-ops-commutable`
Also, it worth cheking that addmod, mulmod work as expected

File: llvm/lib/Target/EVM/EVMOptimizedCodeTransform.cpp [hunk 5]
```diff
@@ -248,39 +253,92 @@ void EVMOptimizedCodeTransform::createStackLayout(Stack TargetStack) {
...
+    // Get the stack layout before the instruction.
+    const Stack &DefaultTargetStack = Layout.operationEntryLayout.at(&Op);
+    size_t DefaultCost =
+        EvaluateStackTransform(CurrentStack, DefaultTargetStack);
+
+    // Commutable operands always take top two stack slots.
+    const unsigned OpIdx1 = 0, OpIdx2 = 1;
+    // Swap the commutable stack items and measure the stack shuffling cost
+    // again.
+    assert(DefaultTargetStack.size() > OpIdx1 &&
```
Comment (akiramenai) [inline llvm/lib/Target/EVM/EVMOptimizedCodeTransform.cpp]: Can it be just `DefaultTargetStack.size() > 1`?

**Review by akiramenai — CHANGES_REQUESTED** [review]
Some minor issues are to be addressed.

File: llvm/lib/Target/EVM/EVMControlFlowGraphBuilder.cpp (hunk 2 above)
Comment (PavelKopyl) [reply llvm/lib/Target/EVM/EVMControlFlowGraphBuilder.cpp]: Yep, thank you!

File: llvm/lib/Target/EVM/EVMOptimizedCodeTransform.cpp (hunk 3 above)
Comment (PavelKopyl) [reply llvm/lib/Target/EVM/EVMOptimizedCodeTransform.cpp:130]: Hm.. but there we check elements of both containers TargetStack and SourceStack?

File: llvm/test/CodeGen/EVM/commutable-ops.ll (hunk 4 above)
Comment (PavelKopyl) [reply llvm/test/CodeGen/EVM/commutable-ops.ll]: Done.

File: llvm/lib/Target/EVM/EVMOptimizedCodeTransform.cpp (hunk 5 above)
Comment (PavelKopyl) [reply llvm/lib/Target/EVM/EVMOptimizedCodeTransform.cpp]: Thank you, fixed.

File: llvm/lib/Target/EVM/EVMOptimizedCodeTransform.cpp (hunk 3 above)
Comment (akiramenai) [reply llvm/lib/Target/EVM/EVMOptimizedCodeTransform.cpp:130]: Sorry, overlooked it.
It might still preferable to use `llvm::zip` in that case as algorithms express intent better.
But if you think it's overly compicated, let's at least not recompute size on each iteration. It tells a reader that it could be changed.

**Review by akiramenai — APPROVED** [review]
LGTM assuming benchmarks show improvment.

File: llvm/lib/Target/EVM/EVMControlFlowGraphBuilder.cpp (hunk 2 above)
Comment (vladimirradosavljevic) [reply llvm/lib/Target/EVM/EVMControlFlowGraphBuilder.cpp]: Sorry for jumping in, but is there any reason why we didn't mark `ADDMOD` and `MULMOD` as commutable in `EVMInstrInfo.td`?
Please note that this is not a blocker for this PR, just curiosity.

Comment (PavelKopyl) [reply llvm/lib/Target/EVM/EVMControlFlowGraphBuilder.cpp]: Thank you for pointing this out. I didn't put `isCommutable` on these instructions, because of non-commutable third argument (module value). I've just I looked at the MachineInstr::isCommutable() internals and other related functions and it looks like we can define them with `isCommutable` flag, because a commutable instruction may have form  `X = op Y, Z, ...`, where only Y and Z operands are commutable.

**Comment by PavelKopyl** [comment]
> LGTM assuming benchmarks show improvment.

The results are following

M3B3:
```
╔═╡ Size (-%) ╞══════════╡ Real life M3B3 ╞═╗
║ Best                                0.375 ║
║ Worst                               0.000 ║
║ Total                               0.196 ║
╠═╡ Cycles (-%) ╞════════╡ Real life M3B3 ╞═╣
║ Best                                0.687 ║
║ Worst                               0.000 ║
║ Total                               0.216 ║
╠═╡ Ergs (-%) ╞══════════╡ Real life M3B3 ╞═╣
║ Best                                0.412 ║
║ Worst                               0.000 ║
║ Total                               0.124 ║
╠══╡ Gas (-%) ╞══════════╡ Real life M3B3 ╞═╣
║ Best                                0.133 ║
║ Worst                               0.000 ║
║ Total                               0.012 ║
╚═══════════════════════════════════════════╝
```

MzB3:
```
╔═╡ Size (-%) ╞══════════╡ Real life MzB3 ╞═╗
║ Best                                0.273 ║
║ Worst                               0.000 ║
║ Total                               0.153 ║
╠═╡ Cycles (-%) ╞════════╡ Real life MzB3 ╞═╣
║ Best                                0.540 ║
║ Worst                               0.000 ║
║ Total                               0.170 ║
╠═╡ Ergs (-%) ╞══════════╡ Real life MzB3 ╞═╣
║ Best                                0.384 ║
║ Worst                               0.000 ║
║ Total                               0.060 ║
╠══╡ Gas (-%) ╞══════════╡ Real life MzB3 ╞═╣
║ Best                                0.107 ║
║ Worst                               0.000 ║
║ Total                               0.011 ║
╚═══════════════════════════════════════════╝

```

File: llvm/lib/Target/EVM/EVMControlFlowGraphBuilder.cpp (hunk 2 above)
Comment (akiramenai) [reply llvm/lib/Target/EVM/EVMControlFlowGraphBuilder.cpp]: Then indeed it's better to use the flag.

**Comment by akiramenai** [comment]
No regressions, as expected. Thank you!

The remaining tasks are to address final comments and squash the commits. I’ll handle the merge afterward.

**Comment by PavelKopyl** [comment]
> No regressions, as expected. Thank you!
> 
> The remaining tasks are to address final comments and squash the commits. I’ll handle the merge afterward.

OK, thank you.

File: llvm/lib/Target/EVM/EVMOptimizedCodeTransform.cpp (hunk 3 above)
Comment (PavelKopyl) [reply llvm/lib/Target/EVM/EVMOptimizedCodeTransform.cpp:130]: Thank you for pointing out, I've used `llvm::zip_equal`.

File: llvm/lib/Target/EVM/EVMControlFlowGraphBuilder.cpp (hunk 2 above)
Comment (PavelKopyl) [reply llvm/lib/Target/EVM/EVMControlFlowGraphBuilder.cpp]: Done.

File: llvm/lib/Target/EVM/EVMOptimizedCodeTransform.cpp (hunk 3 above)
Comment (akiramenai) [reply llvm/lib/Target/EVM/EVMOptimizedCodeTransform.cpp:130]: And you're one step away from `llvm::any`
But let's address it in another review.
